File: sdk/python/virtengine/modules/market.py
"""Market module helpers."""

from __future__ import annotations

import grpc

from virtengine.market.v1beta5 import query_pb2, query_pb2_grpc


class MarketModule:
    """Market module queries."""

    def __init__(self, channel: grpc.aio.Channel):
        self._query = query_pb2_grpc.QueryStub(channel)

    async def orders(self, state: str | None = None) -> query_pb2.QueryOrdersResponse:
        request = query_pb2.QueryOrdersRequest(state=state or "")
        return await self._query.Orders(request)

    async def order(self, order_id: str) -> query_pb2.QueryOrderResponse:
        request = query_pb2.QueryOrderRequest(order_id=order_id)
        return await self._query.Order(request)

    async def leases(self) -> query_pb2.QueryLeasesResponse:
        request = query_pb2.QueryLeasesRequest()
        return await self._query.Leases(request)

    async def lease(self, lease_id: str) -> query_pb2.QueryLeaseResponse:
        request = query_pb2.QueryLeaseRequest(lease_id=lease_id)
        return await self._query.Lease(request)

    async def params(self) -> query_pb2.QueryParamsResponse:
        request = query_pb2.QueryParamsRequest()
        return await self._query.Params(request)

File: sdk/rust/src/modules/market.rs
use tonic::transport::Channel;

use crate::proto::virtengine::market::v1beta5::{
    query_client::QueryClient, QueryLeaseRequest, QueryLeasesRequest, QueryOrderRequest,
    QueryOrdersRequest, QueryParamsRequest, QueryLeaseResponse, QueryLeasesResponse,
    QueryOrderResponse, QueryOrdersResponse, QueryParamsResponse,
};
use crate::{Error, Result};

pub struct MarketModule {
    client: QueryClient<Channel>,
}

impl MarketModule {
    pub fn new(channel: Channel) -> Self {
        Self {
            client: QueryClient::new(channel),
        }
    }

    pub async fn orders(&mut self, state: &str) -> Result<QueryOrdersResponse> {
        let request = QueryOrdersRequest {
            state: state.to_string(),
        };
        self.client
            .orders(request)
            .await
            .map(|res| res.into_inner())
            .map_err(|e| Error::Query(e.to_string()))
    }

    pub async fn order(&mut self, order_id: &str) -> Result<QueryOrderResponse> {
        let request = QueryOrderRequest {
            order_id: order_id.to_string(),
        };
        self.client
            .order(request)
            .await
            .map(|res| res.into_inner())
            .map_err(|e| Error::Query(e.to_string()))
    }

    pub async fn leases(&mut self) -> Result<QueryLeasesResponse> {
        let request = QueryLeasesRequest {};
        self.client
            .leases(request)
            .await
            .map(|res| res.into_inner())
            .map_err(|e| Error::Query(e.to_string()))
    }

    pub async fn lease(&mut self, lease_id: &str) -> Result<QueryLeaseResponse> {
        let request = QueryLeaseRequest {
            lease_id: lease_id.to_string(),
        };
        self.client
            .lease(request)
            .await
            .map(|res| res.into_inner())
            .map_err(|e| Error::Query(e.to_string()))
    }

    pub async fn params(&mut self) -> Result<QueryParamsResponse> {
        let request = QueryParamsRequest {};
        self.client
            .params(request)
            .await
            .map(|res| res.into_inner())
            .map_err(|e| Error::Query(e.to_string()))
    }
}
